examples_tests/35.TAA/taa.frag

```diff
@@ -0,0 +1,61 @@
+#version 430 core
+
+//layout (location = 0) uniform mat4 uPrevVP;
+//layout (location = 1) uniform mat4 uInvVP;
+layout (location = 0) uniform vec2 uJitter;
+
+//layout (binding = 0) uniform sampler2D uDepthBuf;
+layout (binding = 1) uniform sampler2D uCurrentBuf;
+layout (binding = 2) uniform sampler2D uHistoryBuf;
+layout (binding = 3) uniform sampler2D uVelocityBuf;
+
+in vec2 TexCoords;
+
+layout (location = 0) out vec4 outScreen;  //what will be visible on screen
+layout (location = 1) out vec4 outHistory; //history for next frame
+
+float luminanceFromRGB(in vec3 color)
+{
+	return dot(color, vec3(0.299, 0.587, 0.114));
+}
+
+void main()
+{
+/*
+	float depth = texture(uDepthBuf, TexCoords).x; // linearize depth?
+	
+	vec4 clip = vec4(2.0*vec3(TexCoords, depth) - 1.0, 1.0);
+	vec4 world = invVP*clip;
+	world = world/world.w;
+	
+	vec4 q_cs = uPrevVP*world;
+	vec2 q_uv = 0.5*(q_cs.xy/q_cs.w) + 0.5;
+	vec2 vel = TexCoords - q_uv;
+*/
+	// it should sample nearest (depth) in 3x3 neighbourhood
+	vec2 vel = textureLod(uVelocityBuf, TexCoords, 0.0).xy;
+	
+	vec2 q_uv = TexCoords - vel;
+	
+	vec3 current = textureLod(uCurrentBuf, TexCoords, 0.0).rgb;
+	vec3 history = textureLod(uHistoryBuf, q_uv, 0.0).rgb;
+	// AABB clamping, better try (variance?) clipping later (TODO)
+	vec3 neigh0 = textureLodOffset(uCurrentBuf, TexCoords, 0.0, ivec2(1, 0)).rgb;
```

Comment:
use `textureGather()`

Reply:
also these fetches don't make sense, you're getting a cross right now
```
0 x 0
x 0 x
0 x 0
```

do 3 `textureGather`s around the unjittered `TexCoords-uJitter` position, then what you're doing will at least have some meaning/semantic (the gathered samples will be from within output pixel's footprint).

An even higher level of advancement would be to use `texelFetch` and Compute Shader 17x17 shared memory with 16x16 workgroups, then you could get 2x2 or even 4x4 neighbourhoods for the covariance effortlessly.

Reply:
Filmic Temporal SuperSample AA mentions depth-rejection of certain samples from the neighbourhood for a tighter bound and less ghosting (so you'd be using 1-5 samples, instead of 5 all the time for your old method, and 1-16 in my new method).

Reply:
>(the gathered samples will be from within output pixel's footprint)

i dont understand. What is output pixel's footprint?

Reply:
Output pixel `x` and its center + its footprint corners `f`.
```
f 000 f
00000
00x00
00000
f 000 f
```
Jittered pixel center `j` (where sampling, rasterizing and shading occurs)
```
f 000 f
0000 j
00x00
00000
f 000 f
```
If you sample the jittered texture at `j-Jitter` (which is `j-(j-x)==x`) then you're in the 2x2 patch between jittered samples, `textureGather` will get you 4 samples from jittered texture around the unjittered position.

Reply:
Also if you're going to use depth-rejection or object ID to tighten the neighbourhood bounds, then there needs to be a parameter to turn that off (for scenes with translucent media)

Reply:
For transparents, page 46 https://www.dropbox.com/sh/dmye840y307lbpx/AAAQpC0MxMbuOsjm6XmTPgFJa?preview=msalvi_temporal_supersampling.pptx

suggests that we pick the motion vector that is essentially "the most visible" in the neighbourhood (which I guess wouldn't work as well anymore if all objects are equally opaque, but one could make a heuristic so the motion vector of most-contributing blend is picked for motion blur, and then those vectors are dilated with the whole min-depth filter).

Reply:
look at pages 50-53 where the Variance Clipping (with 7x7 gather, not 3x3) makes Stochastic Transparency work.